### ISSUE TITLE
[Feature]: Reach agreement on temporal alignment method/argument names

### ISSUE THREAD

**Issue by CodyCBakerPhD** [catalystneuro/neuroconv#301]
### What would you like to see added to NeuroConv?

From discussions following #237, the usage of the word `timestamps` is a bit overloaded.

Specifically, `timestamps` ought to solely refer to the time vector of the associated `TimeSeries` of each interface, but it currently also applies to the time vectors used in `align_by_interpolation` which may be parsed from TTL or other digital signals.

Also, specific to `align_by_interpolation`, a description of what the function does in words could be "A digital signal is sent from a secondary system and received by the primary system to align the timing differences between the two clocks."

The keyword arguments for the function should be chosen to best describe the role of the inputs

Current: `unaligned_timestamps, aligned_timestamps`
Alternative #1: `ttl_sent_times, ttl_recieved times`
Alternative #2: `unaligned_times, aligned_times`
Alternative #3: `sent_times, recieved_times`
Alternative #3: `secondary_times, primary_times`

For alternative #2, keep in mind that the timing information does not need to be from TTL pulses - it could be inferred from an algorithm similar to how MICrONS does it - or it could be parsed from a non-TTL digital signal, such as audio events

### Do you have any interest in helping implement the feature?

Yes.

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/catalystneuro/neuroconv/blob/main/.github/CODE_OF_CONDUCT.rst)
- [X] Have you ensured this bug was not already [reported](https://github.com/catalystneuro/neuroconv/issues)?

**Comment by CodyCBakerPhD** [comment]
From meeting, this is fine for now

**Comment by CodyCBakerPhD** [comment]
> you claim that align for setters is a good name because the only way to align timestamps is to actually set them.

Not entirely, and I entirely embrace the core logic of your argument

My reasoning is that _a sequence of code written by the user_ would be read by anyone looking at the code or running it _as an affirmation that the timestamps are now aligned_ after calling these methods (or that the motor in the car is repaired or that the dinner is cooked) - of course, this does not mean that they are [in any sense, perhaps it was an identity mapping `interface.align_timestamps(aligned_timestamps=interface.get_timestamps())`], or that they are set correctly (which is something only the experimenter would actually know)

The method and argument names are simply there to communicate the role in the workflow - the goal of which is the align timing information - not necessarily to communicate what exactly that function does or how it works

> For example...

Now, your examples have key differences from the workflow here, so I'll introduce a class-based structure

Let's say you have a `Car` class, and it has an attribute for each part that needs to work in order to successfully call a function called `drive`; my instance of a `car = Car()` cannot call `car.drive()` because `car.motor_working: bool = False`; to fix it, I need to call an intermediary `util` (or more commonly a sequence of several functions ) to repair the motor, then re-attach it (since I can't work on the motor while it's in the car; I had to take it out first)

```python
from car_package.utils import fix_broken_motor

car = Car()

broken_motor = car.motor
repaired_motor = fix_broken_motor(broken_motor=broken_motor)

car.repair_motor(repaired_motor=repaired_motor )
```

Could the method `repair_motor` be called `mount_motor` or `attach_motor`? Absolutely. Does the code above convey that the car now has a repaired motor? I hope so; but let's consider the alternate naming and a user that doesn't name their variables well

```python
from car_package.utils import fix_broken_motor

car = Car()

mtr = np.load("some_motor_file.npy")

car.attach_motor(motor=mtr )
```

is it clear from this that the motor has been fixed? Not without further introspection; but from the method `repair_motor` and keyword `repaired_motor`, whatever the name of the variable you happened to set acts as an affirmation that it is no longer broken.

For completeness, here's the other example adapted as closely to the approximated NeuroConv workflow as possible. I've replaced the role of aligned timestamps with preparation of ingredients, since it is a pre-step of actually cooking the meal

```python
from cooking_package.utils import prepare_ingredients_one_way, prepare_ingredients_another_way

todays_dinner = SpaghettiDinner()  # The `cook_meal` method here takes set prepared ingredients and makes spaghetti out of them

todays_dinner.ingredients = [
    # list of items
]

todays_dinner.ready_to_cook
>>> False
>>> UserWarning: You have not yet prepared your ingredients! Please prepare them ingredients and set the attribute with `.cook_meal(cooked_meal=...)`


prepared_ingredients = prepare_ingredients_another_way(ingredients=todays_dinner.ingredients)
todays_dinner.prepare_ingredients(prepared_ingredients=prepared_ingredients)

cooked_spaghetti = todays_dinner.cook_meal()  # the parellel being `nwbfile = interface.run_conversion()`
```

Now, should this be called `set_prepared_ingredients` for extra clarity? Possibly, probably, IDK. Is it clear and readable enough? I hope


> Now, do I care about this just because I really like logic? No, I think that in the future and for certain simple cases we can actually have a function that fully will deserve the name align_timestamps that gets the timestamps from the right place and use your setters to set the aligned timestamps. If we keep the current names then we overload the concept in advance for no good reason as far as I can see.

Now, your last point here is something we considered early on (particularly, a method called something like `interface.set_alignment_method(alignment_method: callable, **alignment_method_kwargs)` where the user specifies a generic function for performing alignment, and that is what get's called when you actually perform alignment with `align_timestamps`

However, there's such diversity observed across alignment strategies that it was hard to design a consistent enough API on an arbitrary callable to ensure the various pieces worked correctly

Likewise, I do not expect that it will ever be simple enough in the future that each interface can simply have an `align_timestamps` method that somehow knows enough to 'just be able to do it' (automatically align, that is), which is why a setter is all we really can do at the interface level

**Comment by CodyCBakerPhD** [comment]
TL;DR

@h-mayorquin does `set_aligned_timestamps` and `set_aligned_starting_times` work for you?

**Comment by h-mayorquin** [comment]
I think that the crux of our disagreem is this:

```
The method and argument names are simply there to communicate the role in the workflow - the goal of which is the align timing information - not necessarily to communicate what exactly that function does or how it works
```
I disagree with this. I do believe that function names specially when used as verbs should convey as precisely as possible what the function does. 

Instead we are using here something less clear which I can't fully describe. Maybe using the name of the function to "mark a point on the script" (the point at which the alignment should be done) or "remind the code writer what they should be doing with the method"? You will be able to describe it better what is 

To illustrate this difference:
```
from car_package.utils import fix_broken_motor

car = Car()

mtr = np.load("some_motor_file.npy")

car.attach_motor(motor=mtr )
``` 

I understand that you feel that if the method was called `repair_motor` instead of `attach_motor` the code will be clearar. I don't agree with this. The role of the function name is to describe what it does not to remind the author what they should be doing with it. Here, it is the reposnability of the author to provide the clarity with:

```
from car_package.utils import fix_broken_motor

car = Car()

a_working_motor = np.load("some_motor_file.npy")

car.attach_motor(motor=a_working_motor )
``` 

If the me

**Comment by CodyCBakerPhD** [comment]
@h-mayorquin what about `set_aligned_timestamps`? Technically `interface.align_by_interpolation` DOES perform the alignment operation. And now that I double-check it, `align_starting_time` does align the timestamps by the `starting_time`, which is a type of operation in itself

**Comment by CodyCBakerPhD** [comment]
@h-mayorquin Only thing is then you have

`get_original_timestamps`
`get_timestamps`
`set_aligned_timestamps`
`align_starting_time`
`align_by_interpolation`

so `set_` is the odd one out

**Comment by CodyCBakerPhD** [comment]
But it is the only one that is a true setter, that does not perform any other operation

**Comment by CodyCBakerPhD** [comment]
Regardless, my current suggestion for improvement is in https://github.com/catalystneuro/neuroconv/issues/301#issuecomment-1579220671

@h-mayorquin What would _you_ prefer to name each of these methods and their keyword arguments?

**Comment by h-mayorquin** [comment]
Hey, @CodyCBakerPhD , sorry, got into something. First, thanks for bearing with me and laying your arguments so I can pin point more clearly what I think.

First, concerning `car.repair_motor(repaired_motor=a_broken_motor)`,  I do agree with you that this version is unlikely to be used in a misleading form. While I feel that most users don't use keywords to write their functions you could make this keyword only. It stiill sounds strange to me to have an interface that looks like `array.sort_array(sorted_array=the_array_that_you_sorted_somewhere_else)` but point taken that keyword arguments can be used to improve clarity and make the misleading use case of this pattern less likely.

In the same vein I do think that `set_aligned_timestamps` is an improvement as the name more clearly implies that the timestamps should be aligned already. Plus with an appropiatley named keyword, the likelihood of using a method in a misleading way decreases.

I think this is a compromise that dispells my biggest concerns. Your examples were very  useful for the discussion.

Now, what would I prefer?

1) Interfaces have a property called `timestamps` (or something else) and a property `start_time` which can be accessed after initializing the interface so we can do checks before writing (calling `run_conversion` or the future `write` that we have discussed). Both properties can be modified, they are dynamic.

2)  Interfaces also have a property called `native_timestamps`,  `format_timestamps or something else  and `format_start_time` which provide the timestamps and the start_time (timestamps[0]) of the native format when available. Both properties are ready-only.

3) Setters and getters for the `timestamps` and `start_time` for the single and multi-segment cases. When setting the start_time the timestamps are modifed and vice-versa. The interface as an object is time-aware and keeps that as state.

When we call write or run conversion we have logic to decide on the spot if we write timestamps or rate plus start_time if that is possible (regular timestamps and the interface has a sampling rate). Importantly, what we end up writing in an nwbfile is inependent and decoupled of us calling `set_timestamps` or `set_starting_time` We decouple time-awarness from efficient writing data to the nwbfile.

So:
get_original_timestamps  ->  `get_native_timestamps()`
get_timestamps -> `get_timestamps()`
set_aligned_timestamps -> `set_timestamps()` 
align_starting_time -> `set_start_time()`
align_by_interpolation: I think this is a conventience method to create timestamps to be set, I don't really don't think it should be an interface method. In the current state of the code you are leaving it undefined at a lot of places. 

and for multi-segment:

set_aligned_segment_timestamps -> set_segment_timestamps()
set_aligned_segment_starting_times -> set_segment_starting_time().

These methods names are honest, specific and transparent. They don't require further specification with extra keywords or words to avoid a misleading use case. Setting variables is not aligning, the person doing the conversion (and only them) know what the right timestamps are they set them. I feel that aligning is a metaphor that is more loaded to the TTL case and in most cases you just want to shift by a time (so you will use `set_start_time`) or set the timestamps that you extracted from another file (and use `set_timestamps`). Set timestamps and set_start_time seems clear and general enough to me.

**Comment by CodyCBakerPhD** [comment]
> thanks for bearing with me and laying your arguments so I can pin point more clearly what I think.

Same, I just wish we could have done this months ago before it got this deep lol



> In the same vein I do think that set_aligned_timestamps is an improvement as the name more clearly implies that the timestamps should be aligned already. Plus with an appropiatley named keyword, the likelihood of using a method in a misleading way decreases.
> 
> I think this is a compromise that dispells my biggest concerns. Your examples were very useful for the discussion.

I'll move forward with the compromise then; I do really prefer to have the key word 'align' associated with the methods wherever possible because we are not just a generic API for containing data fields; the entire reason we are even adding this capability is specifically to allow and encourage easier temporal alignment

That and more reasons below


> These methods names are honest, specific and transparent.

And uninformative - they lack the key reason as to _why_ we should care or bother with these fields

Most APIs we use for reading (SI, ROIExtractors, OpenCV, DLC, SLEAP, etc...) all have a generic notion of time; it is only in the context of NeuroConv and NWB and the phrase _alignment_ enters in, and is especially critical to pay attention to

> They don't require further specification with extra keywords or words to avoid a misleading use case. Setting variables is not aligning, the person doing the conversion (and only them) know what the right timestamps are they set them.

Right, so they could very well be used in a way that makes sense only to a user that didn't read the documentation and doesn't understand what it's for; I could see generic references being a gateway to the dreaded trialization of all timestamps (native contiguous vector is retrieved, and each value re-indexed relative to event or other interval times)

That still doesn't mean that including the word 'align' will prevent that, but hopefully it would at least give pause, motivation to read the docstring, and even explore the documentation to ask the question "what does 'alignment' mean?"

> I feel that aligning is a metaphor that is more loaded to the TTL case and in most cases you just want to shift by a time (so you will use set_start_time) or set the timestamps that you extracted from another file (and use set_timestamps).

It's not explicitly tied to TTL, which is why TTL isn't in the name (see very first comment in this issue)

Yet in every situation you would want to modify the timestamps in NeuroConv, it is for the purpose of aligning to a common start time

Most cases you will probably just shift by a scalar start time; that shift is done specifically for the purpose of alignment

The other buzzword used is 'synchronization', but that isn't specific enough as there are those other types of synchronizing that we do not encourage (such as data interpolation)
